### PR TITLE
Remodel ros2controlcli

### PR DESCRIPTION
**Pull request by Karsten1987**
This currently sits on top of #310 but also addresses #341.

The current PR of @v-lopez does unfortunately not pass all its tests on OSX and I've noticed that the CLI package was in a rather wild state. This PR is also addressing their linters and refactors a lot of the node handling. 
It also activates the tests of this package in the GH actions.

### CODE REVIEW

**Comment by Karsten1987** [comment]
I will still have to modify the spawner and unspawner scripts to use the python api within the controller manager to get rid of the python subprocess calls.

**Comment by bmagyar** [comment]
merged #310, rebased this one

File: controller_manager/controller_manager/__init__.py [hunk 1]
```diff
@@ -12,8 +12,101 @@
...
 
-from . import launch_utils
+from controller_manager_msgs.srv import ConfigureController, \
+    ListControllers, ListControllerTypes, ListHardwareInterfaces, \
+    LoadController, ReloadControllerLibraries, SwitchController, UnloadController
+
+import rclpy
+
+
+def service_caller(node, service_name, service_type, request):
```
Comment (bmagyar) [inline controller_manager/controller_manager/__init__.py:22]: I know it is common in the ROS codebase to chuck random code in `__init__.py` but I think we should set a better example. Let's create a new file for this with a name indicative of the contents and only use this file to expose them to the outside.

File: controller_manager/controller_manager/launch_utils.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def generate_load_controller_launch_description(controller_name,
     'unload_on_kill' LaunchArguments and a Node action that runs the controller_manager
     spawner.py node to load and start a controller
 
-    Examples
+    Examples;
```
Comment (bmagyar) [inline controller_manager/controller_manager/launch_utils.py:32]: Typo?
```suggestion
    Examples
```

File: ros2controlcli/ros2controlcli/verb/load_configure_controller.py [hunk 3]
```diff
@@ -31,8 +30,8 @@ def add_arguments(self, parser, cli_name):
         add_controller_mgr_parsers(parser)
 
     def main(self, *, args):
+        print("deprecated warning: Please use either 'load --set_state' or 'set_state'")
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/load_configure_controller.py:33]: probably not the best line to raise this but somewhat relevant and at least it shows the command...

what do you think if we called the `command` for setting the state `set_state` but used `state` as the argument when combined with other commands? aka `ros2 control load joint_state_broadcaster --state start`

File: ros2controlcli/ros2controlcli/verb/load_start_controller.py [hunk 4]
```diff
@@ -31,8 +30,8 @@ def add_arguments(self, parser, cli_name):
...
 
     def main(self, *, args):
+        print("deprecated warning: Please use either 'load --set_state' or 'set_state'")
         response = load_start_controller(args.controller_manager, args.controller_name)
         if not response.ok:
-            print('Error loading and starting controller, check '
-                  'controller_manager logs', file=sys.stderr)
-        return not response.ok
+            return 'Error loading and starting controller, check controller_manager logs'
+        return 'Successfully loaded and started controller {}'.format(args.controller_name)
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/load_start_controller.py:37]:
```suggestion
        return f'Successfully loaded and started controller {args.controller_name}'
```

let's be better than printf :D format strings are supported starting from python3.6, we are well within that range with 3.8.5 on ubuntu20

File: controller_manager/controller_manager/__init__.py [hunk 5]
```diff
@@ -12,8 +12,101 @@
...
+    node.get_logger().debug('requester: making request: %r\n' % request)
+    future = cli.call_async(request)
+    rclpy.spin_until_future_complete(node, future)
+    if future.result() is not None:
+        return future.result()
+    else:
+        raise RuntimeError('Exception while calling service: %r' % future.exception())
+
+
+def configure_controller(node, controller_manager_name, controller_name):
```
Comment (bmagyar) [inline controller_manager/controller_manager/__init__.py:40]: finally figured out what puzzled me with the balance of this PR, you didn't remove these calls from `ros2controlcli/ros2controlcli/api/__init__.py` but simply duplicated 'em here. Please remove them from there, if you want to keep """"API compatibility"""" (I'm fairly sure nobody uses this) you can still export these calls from that `__init__.py` too

File: ros2controlcli/ros2controlcli/verb/list_hardware_interfaces.py [hunk 6]
```diff
@@ -25,17 +28,18 @@ def add_arguments(self, parser, cli_name):
...
-        return 0
+        with NodeStrategy(args) as node:
+            hardware_interfaces = list_hardware_interfaces(node, args.controller_manager)
+            command_interfaces = sorted(
+                    hardware_interfaces.command_interfaces, key=lambda hwi: hwi.name)
+            state_interfaces = sorted(
+                    hardware_interfaces.state_interfaces, key=lambda hwi: hwi.name)
+            print('command interfaces')
+            for command_interface in command_interfaces:
+                print('\t%s [%s]' %
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/list_hardware_interfaces.py:39]: let's try always using format strings (aka  `f".... {bla}..."`) and if that doesn't fly (likely in this case) then use `... {}.format()`, it'd be good to get rid of as much hieroglyphs as possible (`%` in this case)

File: ros2controlcli/ros2controlcli/verb/reload_libraries.py [hunk 7]
```diff
@@ -30,9 +32,9 @@ def add_arguments(self, parser, cli_name):
...
-        if response.ok:
-            print('Reload successful')
-        else:
-            print('Error reloading libraries, check controller_manager logs', file=sys.stderr)
-        return not response.ok
+        with NodeStrategy(args) as node:
+            response = reload_controller_libraries(
+                node, args.controller_manager, force_kill=args.force_kill)
+            if response.ok:
+                return 'Reload successful'
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/reload_libraries.py:39]: shouldn't we keep the behaviour of stderr prints when something went wrong?

File: ros2controlcli/ros2controlcli/verb/set_controller_state.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+from controller_manager import configure_controller, list_controllers, switch_controllers
+
+from ros2cli.node.direct import add_arguments
+from ros2cli.node.strategy import NodeStrategy
+from ros2cli.verb import VerbExtension
+
+from ros2controlcli.api import add_controller_mgr_parsers, LoadedControllerNameCompleter
+
+
+class SetControllerStateVerb(VerbExtension):
```
Comment (bmagyar) [inline ros2controlcli/ros2controlcli/verb/set_controller_state.py:24]: index

**Review by bmagyar — CHANGES_REQUESTED** [review]
Good start but I think this PR needs the removal of those subprocess calls too to be complete.
Thanks for adding the gh actions, good catch!

Please always elaborate on OSX issues you find as you are the only source of this info in our group. It'd be nice if you could actually open an issue with them whenever you find some and mark PRs that resolve them, that way we know they are solved for everyone.

File: controller_manager/controller_manager/launch_utils.py (hunk 2 above)
Comment (Karsten1987) [reply controller_manager/controller_manager/launch_utils.py:32]: pep257 wants this.
```
11: ./controller_manager/launch_utils.py:25 in public function `generate_load_controller_launch_description`: D416: Section name should end with a semicolon ('Examples:', not 'Examples')
```

File: ros2controlcli/ros2controlcli/verb/list_hardware_interfaces.py (hunk 6 above)
Comment (Karsten1987) [reply ros2controlcli/ros2controlcli/verb/list_hardware_interfaces.py:39]: the rest of the ros2cli tools are using the same style. We can make this a follow up issue, but I won't address it in this PR.

File: ros2controlcli/ros2controlcli/verb/load_configure_controller.py (hunk 3 above)
Comment (Karsten1987) [reply ros2controlcli/ros2controlcli/verb/load_configure_controller.py:33]: > probably not the best line to raise this but somewhat relevant and at least it shows the command...

fair, but what's a better suited place for it?


I'll change it to `--state`.

File: ros2controlcli/ros2controlcli/verb/reload_libraries.py (hunk 7 above)
Comment (Karsten1987) [reply ros2controlcli/ros2controlcli/verb/reload_libraries.py:39]: the ros2cli infrastructure takes care of this. Any non-zero output will be printed to command line.

File: controller_manager/test/test_spawner_unspawner.cpp [hunk 9]
```diff
@@ -71,7 +71,9 @@ class TestLoadController : public ControllerManagerFixture
 int call_spawner(const std::string extra_args)
 {
   std::string spawner_script = "ros2 run controller_manager spawner.py ";
-  return std::system((spawner_script + extra_args).c_str());
+  auto ret = std::system((spawner_script + extra_args).c_str());
```
Comment (bmagyar) [inline controller_manager/test/test_spawner_unspawner.cpp:74]: debug prints

Comment (Karsten1987) [reply controller_manager/test/test_spawner_unspawner.cpp]: done in 462c736

File: ros2controlcli/ros2controlcli/verb/load_start_controller.py (hunk 4 above)
Comment (Karsten1987) [reply ros2controlcli/ros2controlcli/verb/load_start_controller.py:37]: https://github.com/ros-controls/ros2_control/issues/355

**Comment by Karsten1987** [comment]
So I ended up changing the spawner scripts as well and with these changes the tests seem to work again.

I do believe there's a fine difference in the pep257 between the ros-tooling and ros-industrial CI. The source build (ros-tooling) seems to use a version which expects a semicolon `;` after a doc section, whereas the binary jobs (ros-industrial) don't.
Not sure what to do here.

**Review by bmagyar — APPROVED** [review]

File: ros2controlcli/ros2controlcli/verb/load_configure_controller.py [hunk 10]
```diff
@@ -31,8 +30,8 @@ def add_arguments(self, parser, cli_name):
         add_controller_mgr_parsers(parser)
 
     def main(self, *, args):
+        print("deprecated warning: Please use either 'load --state' or 'set_state'")
```
Comment (destogl) [inline ros2controlcli/ros2controlcli/verb/load_configure_controller.py:33]: `set_state` should be `set_controller_state`

**Comment by destogl** [comment]
@Karsten1987 great job! Thanks for doing this!

(I just opened a few follow-ups to clean the code even more)

File: ros2controlcli/ros2controlcli/verb/load_configure_controller.py (hunk 10 above)
Comment (destogl) [inline ros2controlcli/ros2controlcli/verb/load_configure_controller.py:33]: Remove this in galactic+

**Pull request merged**